staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go

```diff
@@ -199,8 +199,7 @@ func (u *udsGRPCConnector) connect() (proxier, error) {
 		return c, err
 	})
 
-	ctx := context.TODO()
-	tunnel, err := client.CreateSingleUseGrpcTunnel(ctx, udsName, dialOption, grpc.WithInsecure())
+	tunnel, err := client.CreateSingleUseGrpcTunnel(ctx, udsName, dialOption, grpc.WithInsecure(), grpc.WithBlock())
```

Comment:
Removing the `grpc.WithBlock()` and just passing the right context might be enough, but I _think_ we expect the dialer to block on new connections here anyways. @cheftako to help confirm

Reply:
Spoke with @cheftako offline, based on discussions in https://github.com/kubernetes/kubernetes/pull/103272, we're not sure yet if it's safe to pass context into CreateSingleUseGrpcTunnel because it may be shared by multiple requests. See  https://github.com/kubernetes/kubernetes/pull/103272/files#r669878395.

Reply:
The other implementations of connect() use either net.Dial() or tls.Dial() to do the same thing in their implementations. These are both blocking calls. So it seems clear we are expecting connect() to block until it has a function proxier/connection.

Reply:
so this solves the issue of opening multiple connections concurrently, but not the issue of blocking forever?

Reply:
is the problematic goroutine created inside `CreateSingleUseGrpcTunnel` or after we return from here?

Reply:
> is the problematic goroutine created inside CreateSingleUseGrpcTunnel or after we return from here?

Yeah the problematic goroutines are created inside CreateSingleUseGrpcTunnel, from grpc.ClientConn. Here's a sample goroutine pprof I took of an apiserver exhibiting the leak: 

![kube-apiserver-pprof-sample](https://user-images.githubusercontent.com/12699319/155202569-3944cc34-7419-46fe-97a0-0ccb0ceb87ff.png)

From the pprof sample, it seems like there are 4 goroutines leaking every time CreateSingleUseGrpcTunnel is called and when the socket is unavailable:
1. [addrConn.resetTransport()](https://github.com/kubernetes/kubernetes/blob/b631974d68ac5045e076c86a5c66fba6f128dc72/vendor/google.golang.org/grpc/clientconn.go#L815)
2. [ccBalancerWrapper.watcher()](https://github.com/kubernetes/kubernetes/blob/b631974d68ac5045e076c86a5c66fba6f128dc72/vendor/google.golang.org/grpc/balancer_conn_wrappers.go#L67-L99)
3. [transport.reader()](https://github.com/kubernetes/kubernetes/blob/b631974d68ac5045e076c86a5c66fba6f128dc72/vendor/google.golang.org/grpc/internal/transport/http2_client.go#L300)
4. [controlBuffer.get()](https://github.com/kubernetes/kubernetes/blob/b631974d68ac5045e076c86a5c66fba6f128dc72/vendor/google.golang.org/grpc/internal/transport/controlbuf.go#L367-L402)

I _think_ 2/3/4 can be ignored as a side-effect of creating the transport from 1).

Digging through the code, I think the leak is happening because:
1. resetTransport() is running in the background for every call to CreateSingleUseGrpcTunnel()
2. resetTransport() calls tryAllAddrs(), which calls [createTransport()](https://github.com/kubernetes/kubernetes/blob/b631974d68ac5045e076c86a5c66fba6f128dc72/vendor/google.golang.org/grpc/clientconn.go#L1223).
3. createTransport() on error will retry forever since we just passed in context.TODO() in CreateSingleUseGrpcTunnel().
4. Since createTransport() is retrying forever, the transport may eventually be established when proxy server is available again, but it's not guarenteed that apiserver will even use the tunnel returned.   

> so this solves the issue of opening multiple connections concurrently, but not the issue of blocking forever?

Yeah, this could still block forever in theory and is only trying to prevent apiserver from creating spinning up goroutines from CreateSingleUseGrpcTunnel that will retry forever.

Reply:
After digging through the konnectivity-client code a bit more, I found that we're not closing connections if grpcClient.Proxy() errors in CreateSingleUseGrpcTunnel(). I opened https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/328 to fix that specifically and tested that restarting konnectivity proxy does not result in goroutines leaking. The unit test in that PR reproducing the goroutine leak (see https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/328/files#r812506917) is consistent with the pprof data collected as well.

It seems like using `grpc.WithBlock()` masks this issue because it made it very unlikely for `grpcClient.Proxy()` in `CreateSingleUseGrpcTunnel()` to return an error. There's still the question of whether dialing the tunnel should be blocking though given it is the expected behavior of the other tunnel modes. 

Either way I'll open a separate PR for bumping konnectivity-client once that PR is merged.